### ISSUE TITLE
DirectFileStore:  undefined method `unpack' for nil:NilClass

### ISSUE THREAD

**Issue by yosiat** [prometheus/client_ruby#221]
Hi!

I am using this library to expose metrics (in puma server, 2 worker processes) and getting this error when calling `observe` on histogram -
```
/GEM_ROOT/gems/prometheus-client-2.1.0/lib/prometheus/client/data_stores/direct_file_store.rb:241:in initialize
/GEM_ROOT/gems/prometheus-client-2.1.0/lib/prometheus/client/data_stores/direct_file_store.rb:175:in new
/GEM_ROOT/gems/prometheus-client-2.1.0/lib/prometheus/client/data_stores/direct_file_store.rb:175:in internal_store
/GEM_ROOT/gems/prometheus-client-2.1.0/lib/prometheus/client/data_stores/direct_file_store.rb:96:in block in synchronize
/GEM_ROOT/gems/prometheus-client-2.1.0/lib/prometheus/client/data_stores/direct_file_store.rb:161:in block in in_process_sync
/GEM_ROOT/gems/prometheus-client-2.1.0/lib/prometheus/client/data_stores/direct_file_store.rb:161:in synchronize
/GEM_ROOT/gems/prometheus-client-2.1.0/lib/prometheus/client/data_stores/direct_file_store.rb:161:in in_process_sync
/GEM_ROOT/gems/prometheus-client-2.1.0/lib/prometheus/client/data_stores/direct_file_store.rb:95:in synchronize
/GEM_ROOT/gems/prometheus-client-2.1.0/lib/prometheus/client/histogram.rb:69:in observe
```

The issue happens rarely, we are using pretty standard configuration of:
- Rails - latest version of 6
- Puma - 2 worker processes
- Ruby 2.7

In our rails app, I have an initializer which does the following - 
```ruby
# frozen_string_literal: true

require "prometheus/client"
require "prometheus/client/data_stores/direct_file_store"

directory = Rails.root.join("tmp", "metrics").to_s
# Removing metrics directory,  since it can get big on development machines
FileUtils.remove_dir(directory, true)

# initialize DirectFileStore
Prometheus::Client.config.data_store = Prometheus::Client::DataStores::DirectFileStore.new(dir: directory)
```

Any ideas why this issue can happen? Any more information I can try to give to help understand the issue better?

**Comment by dmagliola** [comment]
Hello yosiat, sorry for the very late response...
At first sight, this smells like a race condition to me...

You mentioned Puma has 2 worker processes... Is that 2 worker processes, or 2 threads?
And if processes, how many threads per process?

In addition to that, the initializer code you provided, does that run before Puma forks, or after?
If that runs after forking, there's a possibility that one process will start observing metrics, and then a second one will start up and delete the first one's files.
If that is running before fork, however, I believe that's doing the right thing.

**Comment by yosiat** [comment]
Hi

- 2 processed, each have 32 threads
- It runs before puma forks
- Checked if this happens after worker did a restart and there is no correlation their

when it happens again, I'll try to copy the metrics directory, maybe it will give us some information to debug

**Comment by dmagliola** [comment]
Hello,
Given that you are running in multiple threads, I'm suspecting this may indeed be a race condition.
I've been looking at the code over and over and everything that looks like it needs to be synchronized seems to be synchronized, so I'm not seeing yet where the problem is.

The metrics directory as you mention would be super helpful in debugging this. In the meantime, I will dust off my "performance bashing" script and try to reproduce this myself...

I can see in the last line of your stack trace that this seems to be happening with `histograms`. Do you have multiple reports of the same problem with stack traces ? Is it always histograms?

This may be relevant because any increment on Histograms touches *two* time series (the bucket and the sum). There's a `synchronize` call around those two, but it may be related to that?

Another thing to note is that, according to the stack trace, this is happening when creating a new `FileMappedDict` inside `internal_store`, which happens the first time a metric is observed in a process, or if the PID changed... So it may be related to PID changes also (store initialized before fork, and observed again after fork).

Is it possible you might be observing metrics before Puma forks? (observing, not declaring)
This is not necessarily a problem (we intend to support that), i'm just trying to figure out what could cause this...

Thank you for your patience!

**Comment by yosiat** [comment]
Hi,

Want to come back with more details :)

It happened on a different server instance which have:
- two processes - master & worker
- 16 threads which are processing messages from Rabbit, using Sneakers gem.

I grabbed the metric and created a script that throws these errors, if that's ok I'll send you the data to mail in your Github profile.

What I see is that for a specific histogram metric, the `@position` hash keys are badly formatted, it contains lots of bytes and Unicode characters.

About your questions, we are only declaring the metrics before forking and not observing AFAIK - we declare the metric as a constant in our file, like - `METRIC = registry.histogram(...)`

Thanks for helping out ~!

Yosi

**Comment by dmagliola** [comment]
Weird question... (A [separate issue](https://github.com/prometheus/client_ruby/issues/225) made me think this may be what's happening)

Are you using `with_labels`?
If you are, it's broken, and I can easily imagine how it'd corrupt files now that I put these two issues together...
If you are not, then we're probably back to a threading issue...

>  if that's ok I'll send you the data to mail in your Github profile.

Yes please! 
dmagliola@crystalgears.com is the best address.

**Comment by yosiat** [comment]
Sent files to your mail.

Yes, we are using with_labels only with histograms :) 
If I change `with_labels(xx: "..").increment` to `.increment(labels: { xx: ".." })` it will be ok, or it's the same issue?

**Comment by Sinjo** [comment]
Closing this on the assumption that it has the same underlying cause as #225, which was fixed in #227 and will be part of the upcoming 3.0.0 release.

@yosiat If you find you're still having issues after that, feel free to reopen this issue!
